[ERTP]: Created Intro to ERTP doc (#97)
Agoric/documentation: main/ertp/ERTP-Introduction.md

# ETRP Introduction

ERTP (*Electronic Rights Transfer Protocol*) is Agoric's token

> standard for transferring tokens and other digital assets in

Reviewer: ```suggestion
standard for digital assets in
```
I think "transferring tokens" is redundant

> JavaScript. Using the ETRP API, you can easily create digital assets,

Reviewer: ```suggestion
JavaScript. Using ERTP, you can easily create digital assets,
```
Small typo, API doesn't feel right to me.

> all of which are transferred exactly the same way and with exactly the

Reviewer: I think that this "transferred exactly in the same way..." part doesn't make sense unless you make it clear to the user that a wide variety of digital assets can be represented this way. If the user thinks that only a narrow subset of assets (like ERC20 tokens) can be represented, they will think the fact that the assets are treated in the same way is a given, and this is a weird point.

same security properties. 

ERTP uses object capabilities to enforce access control. If your
program has a reference to an object, it can call methods on that
object. If it doesn't have a reference, it can't. 

## Creating assets with ERTP

In ERTP, *mints* create digital *assets*. Access to an asset type's
mint lets you create more digital assets of that type. You can then
store new assets in a *payment* or a *purse*. 

> - **Payments**: Assets you intend to trade with someone else.

Reviewer: ```suggestion
- **Payments**: Assets you intend to send to someone else.
```
They don't have to be traded, although that's probably a likely case. If we only say trade, it sounds like we're excluding the use case that people would expect.

> - **Purses**: Store assets until you use them.

Reviewer 2: "Store assets until you withdraw them into a payment for use"

> - **Purses**: Store assets until you use them for a trade.

Reviewer: Same thing with trade here.

To send assets in ERTP:
1. Withdraw them from a purse.

> 2. Put them in a payment.

Reviewer: Technically, 1 and 2 are the same step. Do we want them as separate steps for explanatory purposes?

3. Send them to a recipient object as a message.

To receive assets in ERTP:
1. Create a purse for the asset type you'll receive.

> 2. Get access to the asset type you'll receive. (**Note:** you

Reviewer: I think 2 here is not right. The only way you create a purse is by having access to the asset type's issuer. So I think the steps should be:

```suggestion
1. Get access to the issuer for the asset type you expect to receive
2. Create a purse for that asset type
3. Receive the message with the payment and deposit the payment in your purse.
```

> do not need access to the type's mint, just the specific asset you'll 

Reviewer 2: "You do not need access to the type's mint. The Issuer (which is widely shared) will provide an empty purse for the asking."

> do not need access to the type's mint, just the specifc asset you'll 

Reviewer 3: ```suggestion
do not need access to the type's mint, just the speciifc asset you'll 
```

receive. 
3. Receive the message with the payment and put the payment in
your purse.

## Security properties

ERTP purse objects have a `deposit` message which takes a payment
object as its argument. It first checks that the payment object is

> genuine and the same assay type as the purse takes (after all,

Reviewer: ```suggestion
genuine and the same asset type as the purse (after all,
```

quatloos are a very unstable asset, so you don't want to swap your

> very stable simoleans for them). 

Reviewer 2: This mixes exchanging one asset type for another with the concept of fraudulently or accidentally depositing an asset in the wrong kind of purse. Rather than saying "You don't want", I'd say that "We don't want", since this is a security property rather than about not losing value.

`It first checks that the payment object is genuine and the same asset type as the purse (after all, quatloos are a very unstable asset, so we wouldn't want to let people deposit quatloos and get credit for simoleans.)`

> very stable simoleans for them). . 

Reviewer 3: ```suggestion
very stable simoleans for them). 
```

If everything passes the checks, the asset moves from the payment to

> the purse. If there's a problem, it throws an error.

Reviewer: ```suggestion
the purse. If there's a problem, an error is thrown.
```

Or 

```suggestion
the purse. If there's a problem, the process throws an error.
```

After a successful deposit, ERTP guarantees:

> - The payment object is empty.

Reviewer: ```suggestion
- The payment object is deleted from the issuer's records.
```

- The purse contains the payment's full content.

When the `deposit` call throws an error (i.e. something went wrong),
ERTP guarantees: 
- The alleged payment is in the same state as before the call.
- The purse is in the same state as before the call.

## Issuers and mints

Other key ERTP objects are:

- **Mints**: Issue new digital assets as a new Payment. Mints only
issue one kind of asset (quatloos, simoleons, moola, etc.). We refer to
that kind as the mint's *Brand*. So if a mint issues quatloos, it's a
quatloo brand mint.  Only mints can issue new digital assets. To mint

> new assets of a particular type, you much have a reference to that

Reviewer 3: ```suggestion
new assets of a particular type, you must have a reference to that
```

> new assets of a particular type, you much have a reference to that

Reviewer: ```suggestion
new assets of a particular type, you must have a reference to that type's mint.
```

type's mint 

> - **Issuers**: Creates empty purses and payments and maps minted

Reviewer 3: ```suggestion
- **Issuers**: Create empty purses and payments and map minted
```

 assets to them when the assets are added or removed. 
 Issuers verify and move digital assets.

An issuer's special admin facet is a Mint, and that Mint and Issuer are
 associated with each other. With a reference to an issuer, you can
 check the validity of a payment in that issuer's mint's assets, as
 well as claim it either as a new payment to yourself or a purse you
 control. 

Issuers (i.e. their references) should be gotten from a trusted source
and then relied upon as the decider of whether an untrusted payment is

> valid  

Reviewer: ```suggestion
valid.
```

**Note**: There is a one-to-one correspondence between a brand and a
  mint, a mint and an issuer, and an issuer and a brand. In other
  words, each mint issues a unique kind of digital asset, say,
  quatloos, and only that kind of asset.

## Amounts

> *Amounts* describe digital assets. Anyone can make one, and they can

Reviewer: We should say something here about how an `amount` describes a digital asset, but there is no value in an amount. It's only a description, like a description of what is up for auction, or a bank account's balance.

 be sent to anyone. They have two parts:

> - **Brand**: The type of digital asset, such as `qualtoos`.

Reviewer 3: ```suggestion
- **Brand**: An unforgeable object identity for the type of digital asset, such as an object that represents quatloos.
```

> - **Brand**: The type of digital asset, such as `qualtoos`.

Reviewer: ```suggestion
- **Brand**: The type of digital asset, such as `quatloos`.
```

- **Extent**: How much/many of the asset. Fungible extents are natural

>   numbers (**tyg todo**: We don't do, say, 45.7

Reviewer 3: No, just natural numbers for now.  In the future, we will probably have a decimal type for accurate representation of an extent of up to a fixed number of digits.

>   numbers (**tyg todo**: We don't do, say, 45.7

Reviewer: Elsewhere in the docs we say: "Natural numbers are used for fungible digital assets such as money because
 rounding issues make floats problematic. All operations should be
done with the smallest whole unit such that the NatMathHelpers never
deals with fractional parts."

^ this might be helpful

  bitcoin?). Non-fungible extents are strings or objects representing
  attributes of the asset (say, a theater ticket's row and seat positions).

Note: *fungible* means any item in a set can be used. For example, for 
change for a dollar, any four quarters work. *Non-fungible* means
specific items in a set must be used. For  example, theater tickets
are not all the same, and it matters if you get third row center or

> second balcony far left  (and affects what you're willing to trade for

Reviewer: ```suggestion
second balcony far left (and affects what you're willing to trade for
```

it). 

## Amount Math

> Issuers must be able to add and withdraw assets from a purse. This

Reviewer: ```suggestion
Issuers must be able to deposit and withdraw assets from a purse. This
```
Not sure if we should stick with the method names
